tests/shouldwork/TopEntity/PortGeneration.hs

```diff
@@ -0,0 +1,200 @@
+    (PortProduct "outTup" [PortName "outint",PortName "outbool"])
+
+topEntity4 :: Signal System (Gadt Int)
+           -> Signal System Single
+           -> Signal System (CF Int Int Int)
+           -> Signal System (CF Bool Int Int)
+           -> Signal System (OF Int Int Int)
+           -> Signal System (OF Bool Int Int)
+           -> Signal System (X Int Int Int)
+           -> Signal System (X Bool Int Int)
+           -> Signal System Single
+topEntity4 = undefined
+makeTopEntity 'topEntity4
+
+expectedTopEntity4 :: TopEntity
+expectedTopEntity4 =
+  Synthesize "topEntity4"
+    [ PortName "gadt"
+    , PortName "s"
+    , PortProduct "cfiii" [PortName "s"]
+    , PortProduct "cfbii" [PortName "s"]
+    , PortProduct "ofiii" [PortName "s"]
+    , PortProduct "ofbii" [PortName "s"]
+    , PortProduct "" [PortName "xiii", PortName "xiii2"]
+    , PortName "xbii"
+    ]
+    (PortName "s")
+
+topEntity5 :: "in1" ::: Signal System SuccessTy
+           -> "ab"     ::: Signal System (Passthrough (Passthrough Simple Simple) Simple)
+           -> "out" ::: Signal System Int
+topEntity5 = undefined
+makeTopEntity 'topEntity5
+
+expectedTopEntity5 :: TopEntity
+expectedTopEntity5 =
+ Synthesize "topEntity5"
+    [ PortProduct "in1" [PortName "one", PortName "s"]
+    , PortProduct "ab" [PortProduct "" [PortProduct "" [PortName "simple1",PortName "simple2"]
+                                       ,PortProduct "" [PortName "simple1",PortName "simple2"]]
+                       ,PortProduct "" [PortName "simple1",PortName "simple2"]]
+    ]
+    (PortName "out")
+
+expectedTopEntity6 :: TopEntity
+expectedTopEntity6 =
+ Synthesize "topEntity6"
+    [ PortProduct "" [ PortName "clk", PortName "rst", PortName "en"]
+    , PortProduct "in1" [PortName "one", PortName "s"]]
+    (PortName "out")
+
+
+topEntity7 :: (HiddenClockResetEnable System)
+           => "in1" ::: Signal System (Vec 3 Int)
+           -> "in2" ::: Signal System (Vec 3 Simple)
+           -> "passthrough" ::: Signal System (Passthrough Single Single)
+           -> "out" ::: Signal System Int
+topEntity7 = undefined
+makeTopEntity 'topEntity7
+
+expectedTopEntity7 :: TopEntity
+expectedTopEntity7 =
+ Synthesize "topEntity7"
+    [ PortProduct "" [PortName "clk", PortName "rst", PortName "en"]
+    , PortName "in1"
+    , PortName "in2"
+    , PortProduct "passthrough" [PortName "s", PortName "s"]
+    ]
+    (PortName "out")
+
+topEntity8 :: (HiddenClockResetEnable System)
+           => "pair" ::: Signal System (Pair Bool)
+           -> "pair" ::: Signal System (Pair Single)
+           -> "out" ::: Signal System Int
+topEntity8 = undefined
+makeTopEntity 'topEntity8
+
+expectedTopEntity8 :: TopEntity
+expectedTopEntity8 =
+ Synthesize "topEntity8"
+    [ PortProduct "" [PortName "clk", PortName "rst", PortName "en"]
+    , PortProduct "pair" [PortName "left", PortName "right"]
+    , PortProduct "pair" [PortProduct "left" [PortName "s"], PortProduct "right" [PortName "s"]]
```

Comment:
This should probably be:

```haskell
    , PortProduct "pair" [PortName "left", PortName "right"]
    , PortProduct "pair" [PortName "left_s", PortName "right_s"]
```

no?

Reply:
- Done: The failure was in `tests/Clash/Tests/TopEntityGeneration.hs` which needed the change you suggest.
- I've removed the "expected" parts from `tests/shouldwork/TopEntity/PortGeneration.hs` as they are unused.